### PR TITLE
[Enhancement](fe exception) write a java annotation to catch throwable from a method and print log

### LINKED ISSUE #17238: [Enhancement] (fe exception) write a java annotation to catch throwable from a method and print log
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

There are some method in doris that may throw exception but doris does not deal with it. For exmple, FrontendServiceImpl the method getTableNames throws exception, but not print it. If BE call this method and failed, we do not know what happened.  And it is very ugly to add try catch in every method. Maybe we could add a java annotation like  @LOG_EXCEPITON

    @LOG_EXCEPTION
    @Override
    public TGetTablesResult getTableNames(TGetTablesParams params) throws TException {
        LOG.debug("get table name request: {}", params);
        TGetTablesResult result = new TGetTablesResult();
        List<String> tablesResult = Lists.newArrayList();

### Solution

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by y3k-inf**
# Proposed changes

Issue Number: close #17238

## Problem summary

1. How it works?
Aspectj is used to implement the aspect function of annotations. During the compilation process, the aspectj-maven-plugin plugin will automatically weave the code with aspect annotations into the generated classes file.
2. When to use to?
When a method wants to add a try catch to save exception information, the LogException annotation can be used. When there is a method that does not allow errors, the NoException annotation can be used.
3. What is the result when adding this annotation?
Use the LogException annotation to automatically capture exceptions into the Log file, and the code can be more concise. Use the NoException annotation to automatically capture the exception to the Log file and exit the program when an exception occurs.
image

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [x] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
Could you please describe:
1. How it works?
2. When to use to?
3. What is the result when adding this annotation?

File: fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    private static final Logger LOG = LogManager.getLogger(LogAspect.class);
+
+    @Pointcut("@annotation(org.apache.doris.common.annotation.ExceptionLog)")
+    public void throwingLogPointCut() {
+    }
+
+    @AfterThrowing(value = "throwingLogPointCut()", throwing = "e")
+    public void exceptionLog(Exception e)  {
+        LOG.error(e.getMessage());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java:40]: Print 2 logs in one line.
And do we need to print stack trace for each exception? Maybe we can use a flag to control it?

**Comment by NetShrimp06** [comment]
> * How it works?
>   Aspectj is used to implement the aspect function of annotations. During compilation, the aspectj-maven-plugin plug-in will automatically weave the code added with aspect annotations into the generated classes file.
> * When to use to?
>   When a method wants to add a try catch to save exception information, this annotation can be used to automatically capture the exception into the Log file, and the code can be more concise.
> * What is the result when adding this annotation?
>   When an exception occurs in the method, the exception information will be recorded in the log file. If you want to print the stack information, you can set the isGetStackTrace parameter of the annotation to true (the default is false to not print the stack information).

File: fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java (hunk 1 above)
Comment (NetShrimp06) [reply fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java:48]: I added a parameter to let the user decide whether to print the stack information, the default is not to print

**Comment by NetShrimp06** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.99 seconds
 stream load tsv:          446 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         37 seconds loaded 2358488459 Bytes, about 60 MB/s
 stream load orc:          73 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230316034752_clickbench_pr_114973.html

File: fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @AfterThrowing(value = "throwingLogPointCut()", throwing = "e")
+    public void exceptionLog(JoinPoint point, Exception e)  {
+        MethodSignature joinPointObject = (MethodSignature) point.getSignature();
+        Method method = joinPointObject.getMethod();
+        ExceptionLog exceptionLog = method.getAnnotation(ExceptionLog.class);
+        LOG.warn(e.getMessage());
+        if (exceptionLog != null && exceptionLog.isGetStackTrace()) {
+            LOG.warn(getStackTrace(e));
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java:50]: LOG.warn(e) will print stack trace automatically.

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/BulkLoadJob.java [hunk 3]
```diff
@@ -181,6 +182,7 @@ public Set<String> getTableNamesForShow() {
                 .collect(Collectors.toSet());
     }
 
+    @ExceptionLog
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/BulkLoadJob.java:185]: using  @LogException instead.

File: fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+public class LogAspect {
+
+    private static final Logger LOG = LogManager.getLogger(LogAspect.class);
+
+    @Pointcut("@annotation(org.apache.doris.common.annotation.ExceptionLog)")
+    public void throwingLogPointCut() {
+    }
+
+    @AfterThrowing(value = "throwingLogPointCut()", throwing = "e")
+    public void exceptionLog(JoinPoint point, Exception e)  {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java:44]: using Throwable intead of Exception,  Exception is only subset to errors.

File: fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.annotation.Pointcut;
+import org.aspectj.lang.reflect.MethodSignature;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.reflect.Method;
+
+@Aspect
+public class LogAspect {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/aspectj/LogAspect.java:35]: Could you please add another annotation. 
@NoException
If a method throws any exceptions, should print log and call system.exit.

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by yiguolei** [comment]
run buildall

**Comment by englefly** [comment]
it does not work on mac, or newer version of JDK

**Pull request merged**